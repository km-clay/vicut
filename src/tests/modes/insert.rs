use crate::tests::{normal_cmd, vicut_integration};

use super::*;

#[test]
fn vimode_insert_structures() {
	let raw = "abcdefghijklmnopqrstuvwxyz1234567890-=[];'<>/\\x1b";
	let mut mode = ViInsert::new();
	let cmds = mode.cmds_from_raw(raw);
	insta::assert_debug_snapshot!(cmds)
}

#[test]
fn two_inserts() {
	vicut_integration(
		"foo bar biz",
		&[ "-m", "iInserting some text<esc>2wiAnd some more here too<esc>" ],

		"Inserting some textfoo bar And some more here toobiz",
	);
}

#[test]
fn ctrl_w() {
	vicut_integration(
		"foo bar biz",
		&[ "-m", "eiInserting_some_text<c-w>" ],

		"foo bar biz",
	);
}

#[test]
fn linebreaks() {
	// Also tests 'a' at the end of the buffer
	vicut_integration(
		"foo bar biz",
		&[ "-m", "$a<enter>bar foo biz" ],
		"foo bar biz\nbar foo biz",
	);
	vicut_integration(
		"foo bar biz",
		&[ "-m", "$a<CR>bar foo biz" ],
		"foo bar biz\nbar foo biz",
	)
}

#[test]
fn navigation() {
	vicut_integration(
		"foo bar biz\nbar foo biz",
		&[
			"-m", "j",
			"-m", "<right><right><right><right><up>",
			"-c", "e",
		],
		"bar",
	)
}

#[test]
fn backspace_and_delete() {
	vicut_integration(
		"foo bar biz\nbar foo biz",
		&[
			"-m", "jw",
"-m", "i<BS><BS><BS><BS><del><del><del><del>",
			"-c", "e"
		],
		"biz",
	)
}

#[test]
fn end_of_line_motion_boundary() {
	vicut_integration(
		"foo bar", 
		&[
			"-m", "$",
			"-m", "i<right>",
			"-c", "b"
		], 
		"bar"
	);
}

#[test]
fn prefix_insert() {
	vicut_integration(
		"    foo bar", 
		&[
			"-m", "$",
			"-m", "Iinserting some text at the start"
		], 
		"    inserting some text at the startfoo bar"
	);
}

#[test]
fn insert_unicode() {
	vicut_integration(
		"foo", 
		&[
			"-m", "ea→bar",
		], 
		"foo→bar"
	);
}

#[test]
fn insert_in_empty_line() {
	vicut_integration(
		"foo\n\nbiz", 
		&[
			"-m", "jibar",
		], 
		"foo\nbar\nbiz"
	);
}

#[test]
fn insert_from_visual_mode() {
	vicut_integration(
		"foo biz bar", 
		&[
			"-m", "wveIinserting some text",
		], 
		"inserting some textfoo biz bar"
	);
	vicut_integration(
		"foo biz bar", 
		&[
			"-m", "wveAinserting some text",
		], 
		"foo bizinserting some text bar"
	);
}

#[test]
fn insert_empty_buffer() {
	vicut_integration(
		"",
		&[
			"-m", "ihello world"
		],
		"hello world"
	);
}
