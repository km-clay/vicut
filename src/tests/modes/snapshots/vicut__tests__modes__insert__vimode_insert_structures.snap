---
source: src/tests/modes/insert.rs
expression: cmds
---
[
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'a',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'b',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'c',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'd',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'e',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'f',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'g',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'h',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'i',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'j',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'k',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'l',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'm',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'n',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'o',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'p',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'q',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'r',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    's',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    't',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'u',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'v',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'w',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'x',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'y',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'z',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '1',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '2',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '3',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '4',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '5',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '6',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '7',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '8',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '9',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '0',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '-',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '=',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '[',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    ']',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    ';',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '\'',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '<',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '>',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '/',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '\\',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'x',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    '1',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
    ViCmd {
        register: RegisterName {
            name: None,
            count: 1,
            append: false,
        },
        verb: Some(
            VerbCmd(
                1,
                InsertChar(
                    'b',
                ),
            ),
        ),
        motion: Some(
            MotionCmd(
                1,
                ForwardChar,
            ),
        ),
        raw_seq: "",
        flags: CmdFlags(
            0x0,
        ),
    },
]
